app/services/smart_proxy_reference.rb

```diff
@@ -0,0 +1,51 @@
+module SmartProxyReference
+  class << self
+    attr_accessor :references
+
+    def smart_proxy_reference(hash)
+      raise Foreman::Exception.new(N_("Proxy reference should be { :relation => [:column_name, ...] }")) unless correct_reference_format?(hash)
+      hash.deep_clone
+    end
+
+    def add_smart_proxy_reference(hash)
+      hash_clone = smart_proxy_reference hash
+      if @references
+        join_references @references, hash_clone
+      else
+        @references = hash_clone
+      end
+    end
+
+    def smart_proxy_references
+      result = local_entries.deep_clone
+      join_references result, entries_from_plugins
+    end
+
+    def local_entries
```

Comment:
why calling `local_entries` the method that returns `referencies` Shouldn't we either rename one or the other to match each other?